daffodil-propgen/src/main/scala/org/apache/daffodil/propGen/TunableGenerator.scala

```diff
@@ -0,0 +1,223 @@
+    |
+    |import org.apache.daffodil.exceptions.Assert
+    |import org.apache.daffodil.schema.annotation.props.EmptyElementParsePolicy
+    |import org.apache.daffodil.util.LogLevel
+    |import org.apache.daffodil.util.Logging
+    |import org.apache.daffodil.util.Misc
+    |import org.apache.daffodil.xml.DaffodilXMLLoader
+    |
+    |object DaffodilTunables {
+    |
+    |  def apply(tunables: Map[String, String]): DaffodilTunables = {
+    |    apply().setTunables(tunables)
+    |  }
+    |
+    |  def apply(tunable: String, value: String): DaffodilTunables = {
+    |    apply().setTunable(tunable, value)
+    |  }
+    |
+    |  def apply(): DaffodilTunables = {
+    |    // override tunables from the global configuration file on the class path, if it exists
+    |    val (configOpt, _) = Misc.getResourceOption("/daffodil-config.xml")
+    |    val configTunables: Map[String, String] =
+    |      if (configOpt.isDefined) {
+    |        val loader = new DaffodilXMLLoader()
+    |        val node = loader.load(URISchemaSource(configOpt.get))
+    |        val trimmed = scala.xml.Utility.trim(node)
+    |        val tunablesNode = (trimmed \ "tunables").headOption
+    |        val tunablesMap: Map[String, String] = tunablesNode match {
+    |          case None => Map.empty
+    |          case Some(tunableNode) => {
+    |            tunableNode.child.map { n => (n.label, n.text) }.toMap
+    |          }
+    |        }
+    |        tunablesMap
+    |      } else {
+    |        Map.empty
+    |      }
+    |
+    |    new DaffodilTunables().setTunables(configTunables)
+    |  }
+    |}
+    |
+    |case class DaffodilTunables private (
+    """.trim.stripMargin
+
+  val middle = """
+    |  extends Serializable
+    |  with Logging {
+    |
+    |  def setTunables(tunables: Map[String, String]): DaffodilTunables = {
+    |    tunables.foldLeft(this) { case (dafTuns, (tunable, value)) => dafTuns.setTunable(tunable, value) }
+    |  }
+    |
+    |  def setTunable(tunable: String, value: String): DaffodilTunables = {
+    |    tunable.toLowerCase match {
+    """.trim.stripMargin
+
+  val bottom = """
+    |      case _ => throw new IllegalArgumentException("Unknown tunable: " + tunable)
+    |    }
+    |  }
+    |
+    |  def notSuppressedWarning(warnID: WarnID) =
+    |    !suppressSchemaDefinitionWarnings.contains(warnID) &&
+    |      !suppressSchemaDefinitionWarnings.contains(WarnID.All)
+    |
+    |}
+    """.trim.stripMargin
+
+  val tunablesRoot = (schema \ "element").find(_ \@ "name" == "tunables").get
+  val tunableNodes = tunablesRoot \\ "all" \ "element"
+
+  def writeGeneratedCode(w: java.io.FileWriter) {
+    val tunables =
+      tunableNodes.map { tunableNode =>
+        val schemaName = tunableNode \@ "name"
+        val schemaType = tunableNode \@ "type"
+        val schemaDefault = tunableNode \@ "default"
+
+        schemaName match {
+          case "defaultEmptyElementParsePolicy" =>
```

Comment:
I am uncertain as to why this case statement? This will require editing for every new property we add.  If this is unavoidable (or unavoidable without additional work you didn't plan to do currently), then a comment to the effect that this has to be edited as well as the dafext.xsd is in order.

Reply:
It's only needed right now for tunables that aren't primitives (i.e. enums or lists of enums). But I guess anything that isn't a primitive can be assumed to be an enum. And then we just need the special caase for suppressSchemaDefinitionWarnings. I"ll make this change.